File: app/src/main/java/com/agrogames/islandsofwar/ui/impl/UI.java
package com.agrogames.islandsofwar.ui.impl;

import android.util.Pair;

import com.agrogames.islandsofwar.graphics.abs.TextureDrawer;
import com.agrogames.islandsofwar.ui.abs.Element;
import com.agrogames.islandsofwar.ui.abs.ElementType;

import java.util.ArrayList;
import java.util.List;

public class UI implements com.agrogames.islandsofwar.ui.abs.UI {
    private final List<Element> elements = new ArrayList<>();
    @Override
    public Element createElement(ElementType type, float x, float y, float width, float height, String texture) {
        Element element = new com.agrogames.islandsofwar.ui.impl.Element();
        element.setType(type);
        element.setX(x);
        element.setY(y);
        element.setWidth(width);
        element.setHeight(height);
        element.setTexture(texture);
        element.setVisible(true);
        element.setRenderInBorders(true);
        elements.add(element);
        return element;
    }

    @Override
    public void deleteElement(Element element) {
        elements.remove(element);
    }

    @Override
    public boolean render(TextureDrawer drawer, Float touchX, Float touchY) {
        for (Element element : elements){
            if(!element.getVisible()) continue;
            if(element.getRenderInBorders()) {
                drawer.drawTexture(element.getX(), element.getY(), element.getTexture(),
                        element.getWidth(), element.getHeight(), 0);
            } else {
                Pair<Float, Float> size =
                        drawer.drawTexture(element.getX(), element.getY(), element.getTexture(), 0);
                if(element.getWidth() == -1 || element.getHeight() == -1) {
                    element.setWidth(size.first);
                    element.setHeight(size.second);
                }
            }
            if(touchX != null && touchY != null){
                if(element.getX() + element.getWidth() / 2 > touchX
                && element.getX() - element.getWidth() / 2 < touchX
                && element.getY() + element.getHeight() / 2 > touchY
                && element.getY() - element.getHeight() / 2 < touchY){
                    if(element.callListener()){
                        return true;
                    }
                }
            }
        }
        return false;
    }
}

File: app/src/main/java/com/agrogames/islandsofwar/ui/impl/UI.kt
package com.agrogames.islandsofwar.ui.impl

import com.agrogames.islandsofwar.engine.abs.common.Point
import com.agrogames.islandsofwar.graphics.impl.drawtexture.TextureDrawer

class UI {
    private val elements: MutableList<Element> = ArrayList()
    fun createElement(element: Element): Element {
        elements.add(element)
        return element
    }

    fun deleteElement(element: Element) {
        elements.remove(element)
    }

    fun render(drawer: TextureDrawer, touch: Point?): Boolean {
        for (element in elements) {
            if (!element.visible) continue
            if (element.renderInBorders) {
                drawer.drawTexture(element.x, element.y, element.texture,
                        element.width, element.height, 0f)
            } else {
                if (element.width == -1f || element.height == -1f) {
                    val size = drawer.drawTexture(element.x, element.y, element.texture, 0f)
                    element.width = size.first
                    element.height = size.second
                } else {
                    val size = drawer.getSize(element.texture)
                    if (size.first > element.width && size.second > element.height) {
                        element.renderInBorders = true
                        drawer.drawTexture(element.x, element.y, element.texture,
                                element.width, element.height, 0f)
                    } else {
                        drawer.drawTexture(element.x, element.y, element.texture, 0f)
                        element.width = size.first
                        element.height = size.second
                    }
                }
            }
            if (touch != null) {
                if (element.x + element.width / 2 > touch.x && element.x - element.width / 2 < touch.x
                    && element.y + element.height / 2 > touch.y && element.y - element.height / 2 < touch.y) {
                    if (element.onClick != null) {
                        element.onClick?.let { it(element) }
                        return true
                    }
                }
            }
        }
        return false
    }
}